libs/libarchfpga/src/physical_types.h

```diff
@@ -62,6 +65,74 @@ struct t_cluster_placement_primitive;
 struct t_arch;
 enum class e_sb_type;
 
+/****************************************************************************/
+/* FPGA metadata types                                                      */
+/****************************************************************************/
+/* t_offset, t_metadata_as, and t_metadata_dict provide a types to store
+ * metadata about the FPGA architecture and routing routing graph along side
+ * the pb_type, grid, node and edge descriptions.
+ *
+ * The metadata is stored as a simple key/value map.  key's are string and an
+ * optional coordinate. t_metadata_as provides the value storage, which is a
+ * string.  t_metadata_as has some convience methods for a handful of types
+ * (e.g. int, vector of int), but the underlying data type is always a string.
+ */
+
+// Metadata value storage.
+class t_metadata_as {
+public:
+    explicit t_metadata_as(std::string v) : value_(v) {}
+    explicit t_metadata_as(const t_metadata_as &o) : value_(o.value_) {}
+
+    // Return string value.
+    std::string as_string() const { return value_; }
+
+private:
+    std::string value_;
+};
+
+// Metadata storage dictionary.
+struct t_metadata_dict : std::unordered_map<
+                            std::string,
+                            std::vector<t_metadata_as>> {
+    // Is this key present in the map?
+    inline bool has(std::string key) const {
+        return this->count(key) >= 1;
+    }
+
+    // Get all metadata values matching key.
+    //
+    // Returns nullptr if key is not found.
+    inline const std::vector<t_metadata_as>* get(std::string key) const {
+        auto iter = this->find(key);
+        if(iter != this->end()) {
+          return &iter->second;
+        }
+        return nullptr;
+    }
+
+    // Get metadata values matching key.
+    //
+    // Returns nullptr if key is not found or if multiple values are prsent
+    // per key.
+    inline const t_metadata_as* one(std::string key) const {
+        auto values = get(key);
+        if (values == nullptr) {
+            return nullptr;
+        }
+        if (values->size() != 1) {
+            return nullptr;
+        }
+        return &((*values)[0]);
+    }
+
+    // Adds value to key.
+    void add(std::string key, std::string value) {
+        auto iter = this->emplace(key, std::vector<t_metadata_as>());
```

Comment:
iter isn't a great name for the variable here; it's really a pair<iterator,bool>

Again, define the behaviour with repeated keys.  Emplace will just return <iterator,false> to the existing entry if the key is already in the unordered_map, and the empty vector will be discarded.  This will append the new value if the key already exists, but you have to read the code pretty carefully to figure this out so a comment defining the behaviour with repeated keys would be useful.

Reply:
The current behavior is intentional and correct.   This code is the idiomatic way to insert if required and get the iterator matching the key.  Anyways, I've renamed the variable and added a comment.